framework/src/modules/dpos_v2/commands/vote_delegate.ts

```diff
@@ -184,18 +238,55 @@ export class VoteDelegateCommand extends BaseCommand {
 				// Sort account.unlocking
 				sortUnlocking(voterData.pendingUnlocks);
 
-				if (voterData.pendingUnlocks.length > MAX_UNLOCKING) {
-					throw new Error(`Pending unlocks cannot exceed ${MAX_UNLOCKING.toString()}.`);
+				if (voterData.pendingUnlocks.length > MAX_NUMBER_PENDING_UNLOCKS) {
+					this.events.get(DelegateVotedEvent).error(
+						context,
+						{
+							senderAddress,
+							delegateAddress: vote.delegateAddress,
+							amount: vote.amount,
+						},
+						PoSEventResult.VOTE_FAILED_TOO_MANY_PENDING_UNLOCKS,
+					);
+
+					throw new Error(
+						`Pending unlocks cannot exceed ${MAX_NUMBER_PENDING_UNLOCKS.toString()}.`,
+					);
+				}
+
+				await this._internalMethod.assignVoteRewards(
+					context,
+					senderAddress,
+					voterData.sentVotes[index],
+					delegateData,
+				);
+
+				// sent vote is not removed from voerData because of zero amount
+				if (aggregateVotedAmount !== BigInt(0)) {
+					voterData.sentVotes[index].amount = aggregateVotedAmount;
+					voterData.sentVotes[index].voteSharingCoefficients = delegateData.sharingCoefficients;
 				}
 			} else {
 				// Upvote amount case
-				const upvote =
-					originalUpvoteIndex > -1
-						? voterData.sentVotes[originalUpvoteIndex]
-						: {
-								delegateAddress: vote.delegateAddress,
-								amount: BigInt(0),
-						  };
+				let upvote;
+
+				if (originalUpvoteIndex > -1) {
+					upvote = voterData.sentVotes[originalUpvoteIndex];
+
+					await this._internalMethod.assignVoteRewards(
+						context.getMethodContext(),
+						senderAddress,
+						voterData.sentVotes[originalUpvoteIndex],
+						delegateData,
```

Comment:
here also it needs to update the sharing coefficient?

Reply:
Isn't this part of `assingVoteRewards`?

Reply:
This change was made based on recommendation from @shuse2 to make assignVoteRewards simpler and to avoid state mutation on arguments. [assignVoteRewards](https://github.com/LiskHQ/lisk-sdk/pull/7777/files#diff-43c606e7d148b35562b0dceffa505c137a8115e0ef5741b4658525321b599f94) calculates, unlock and transfer reward amount for each sharing coefficient of the delegate. To achieve the same result as per the LIP. The assignment of `voteSharingCoefficient` moved out into execution logic.

Reply:
Ok sounds good. Do you want to update the LIP accordingly, or not?

Reply:
Not critical, but probably for clarity, it would be nice to update the LIP

Reply:
Will do that in the current PR. Btw, did you also take this into account in the reward sharing implementation? The function assignVoteRewards is also called in the execution of the claim reward command..

Reply:
i think it just need to follow the same approach of assign reward then update sharing coefficient?

Reply:
Yep! Updated also the LIPs in the open PR in public repo. 

The idea was that since the state change should always happen just after assignVoteRewards, it seemed more natural to me to include it in the function, but it's ok now as well (given that we always remember to do the state change after the call)